instagram behavior: before starting post traversal, view first post t…

ikreymer opened the pull request:
…o save direct post view assets

Instagram loads additional assets for 'single-post view', but is not normally captured during behavior. Fortunately, forcing it to navigate to that view for first post only, enables the view to be usable for all other posts.

ikreymer committed:
instagram behavior: before starting post traversal, view first post to save direct post view assets

N0taN3rd commented on behaviors/instagram/user.js:
```diff
@@ -81,6 +81,26 @@ async function* handlePost(post, { cliAPI, info }) {
   }
 }
 
+function loadPostView(preTraversal) {
```
It would be simplest to do this

```suggestion
function viewStoriesAndLoadPostView(cliAPI, info) {
  return async function* preTraversal() {
    if (shared.loggedIn(cliAPI.$x)) {
      // viewing stories change the markup of the page to be stories mode
      // not timeline mode and reverts back to timeline mode once done #react
      // thus we can only hold a reference to the element that will start
      // a story chain we will view now
      const profilePic = lib.qs('img[alt*="profile picture"]');
      if (
        profilePic &&
        window.getComputedStyle(profilePic).cursor === 'pointer'
      ) {
        yield* shared.viewStories(profilePic, info, true);
      }
      if (lib.selectorExists(selectors.userOpenStories)) {
        yield* shared.viewStories(lib.qs(selectors.userOpenStories), info);
      }
    }

    yield lib.stateWithMsgNoWait('Loading post view', info.state);
    const firstPostHref = lib.qs('article a').href;
    window.history.pushState({}, "", firstPostHref);
    window.dispatchEvent(new PopStateEvent("popstate", {state: {}}));

    await lib.delay(2000);

    window.history.back();

    await lib.delay(2000);
    yield lib.stateWithMsgNoWait('Loaded post view', info.state);
  }
}
```

N0taN3rd commented on behaviors/instagram/user.js:
```diff
@@ -111,6 +131,9 @@ export default function instagramUserBehavior(cliAPI) {
+  preTraversal = loadPostView(preTraversal);
+
   return lib.traverseChildrenOfCustom({
     preTraversal,
```
```suggestion
    preTraversal: viewStoriesAndLoadPostView(cliAPI, info),
```

N0taN3rd reviewed (commented):
The loading of the actual post page always would allow for a simplification

ikreymer committed:
Apply suggestions from code review

preTraversal refactor

Co-Authored-By: John Berlin <n0tan3rd@gmail.com>

ikreymer committed:
finish suggested changes, refactor preTraversal setup

ikreymer committed:
improve load post view logic:
- wait for <article> to change instead of fixed amount
- use replaceState / dispatch popstate or both changes, avoid using window.history for now
